core/src/main/java/org/mule/runtime/core/api/EventContext.java

```diff
@@ -107,4 +108,30 @@
    * @return {@code true} if the source system provided a correlation id, {@code false otherwise}.
    */
   boolean isCorrelationIdFromSource();
+
+  /**
+   * @return An immutable list with all the child context which were produced from {@code this} instance
+   */
+  List<EventContext> getChildContexts();
+
+  /**
+   * Returns {@code this} context's parent if it has one
+   * @return {@code this} context's parent or {@link Optional#empty()} if it doesn't have one
+   */
+  Optional<EventContext> getParentContext();
+
+  /**
+   * @return Whether {@code this} context and all its childs have been completed
+   */
+  boolean isTerminated();
+
+  /**
+   * Indicates that the owning {@link Event} is involved in at least one streaming operation
+   */
+  void streaming();
```

Comment:
you asked for this!!! The idea is that each context knows if it's part of a streaming operation so that we can avoid access to the shared state necessary to properly dispose the buffer.